### PR TITLE
Add replaceTensor to swap floating point type

### PR DESCRIPTION
**Pull request by junjihashimoto**
This PR enables training in half type.
Half type is only supported on cuda.

### CODE REVIEW

File: hasktorch/src/Torch/NN.hs [hunk 1]
```diff
@@ -118,12 +119,12 @@ class Parameterized f where
   default _replaceParameters :: (Generic f, GParameterized (Rep f)) => f -> ParamStream f
   _replaceParameters f = to <$> _gReplaceParameters (from f)
 
-  replaceDevice :: Device -> f -> f
-  default replaceDevice :: (Generic f, GParameterized (Rep f)) => Device -> f -> f
-  replaceDevice dev f = to $ gReplaceDevice dev (from f)
+  replaceTensor :: (Tensor -> Tensor) -> f -> f
```
Comment (tscholak) [inline hasktorch/src/Torch/NN.hs:122]: this is a nice idea. can you make this dependent on the generic representation if `f`, please?
this way, one could use that to do more intentional model weight surgery.

Comment (junjihashimoto) [reply hasktorch/src/Torch/NN.hs:122]: I think there is a demand to match and replace specific data, but I'm not sure how to make it without creating a custom class and instance for each application. I wish there was something like try-function for to-function of ghc.generic.

Comment (junjihashimoto) [reply hasktorch/src/Torch/NN.hs:122]: We can do it with template haskell, but it's not what we expect.

Comment (junjihashimoto) [reply hasktorch/src/Torch/NN.hs:122]: Oh, I misunderstood it.

Comment (junjihashimoto) [reply hasktorch/src/Torch/NN.hs:122]: I misunderstood a little, please wait a moment.

**Comment by junjihashimoto** [comment]
@tscholak I've added GTraversal type class to support for generic replacement.
We will need to change Paramerized instances.

File: hasktorch/src/Torch/Traversable.hs [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+  gflatten' (M1 x) = gflatten' x
+  gupdate' (M1 x) = do
+    x' <- gupdate' x
+    return $ M1 x'
+
+instance GTraversable a a where
+  gflatten = pure
+  gupdate _ = gpop
+
+instance {-# OVERLAPS #-} (Generic f, GGTraversable a (Rep f)) => GTraversable a f
```
Comment (tscholak) [inline hasktorch/src/Torch/Traversable.hs:68]: I'm not sure about this. How is the user experience if
1. No Generic instance exists  for a given f?
2. No Generic instance can be automatically derived for a given f?
3. A user wishes to implement a custom GTraversable instance for a given f?

Comment (junjihashimoto) [reply hasktorch/src/Torch/Traversable.hs:68]: > 1. No Generic instance exists for a given f?

f can become a GTraversable instance without a Generic instance.

> 2. No Generic instance can be automatically derived for a given f?

We can use a custom Generic instance for a given f.

> 3. A user wishes to implement a custom GTraversable instance for a given f?

A user can implement a custom GTraversable instance of f without a Generic instance.
A user can not implement a custom GTraversable instance of f with a Generic instance.

Oh, we can not use both a custom Generic instance and a custom GTraversable instance for a given f.

Comment (junjihashimoto) [reply hasktorch/src/Torch/Traversable.hs:68]: I will make some examples for the instances.

File: hasktorch/test/NNSpec.hs [hunk 3]
```diff
@@ -16,6 +17,67 @@ import Torch.Autograd
...
+  }
+  deriving (Show, Generic)
+
+instance GTraversable Parameter Customlinear where
+  gflatten Customlinear {..} = [customWeight, customBias, customBias] -- These duplicated biases are for testing!!
+  gupdate _ = do
+    customWeight <- gpop
+    customBias <- gpop
+    _ <- gpop
+    return $ Customlinear {..}
```
Comment (junjihashimoto) [inline hasktorch/test/NNSpec.hs:38]: > 1. No Generic instance exists for a given f?
> 2. No Generic instance can be automatically derived for a given f?

We can implement a custom GTraversable instance with or without a generic instance.

> 3. A user wishes to implement a custom GTraversable instance for a given f?

I think so.
I think a user wishes to use a generic GTraversable instance for the time being.

Comment (junjihashimoto) [reply hasktorch/test/NNSpec.hs:38]: An additional general instance is not allowed. 
The following code causes  a duplicated instance error.
```
instance {-# OVERLAPS #-} (Scalar f) => GTraversable a f
```

**Pull request closed without merging**